chain/chain/src/chain.rs

```diff
@@ -5865,22 +5862,66 @@ impl<'a> ChainUpdate<'a> {
         Ok(true)
     }
 
+    /// This function assumes `BlockInfo` is already retrievable from `epoch_manager`.
+    /// This can be achieved by calling `add_validator_proposals`.
+    #[cfg(feature = "new_epoch_sync")]
+    fn save_epoch_sync_info_if_finalised(&mut self, header: &BlockHeader) -> Result<(), Error> {
+        let block_info = self.epoch_manager.get_block_info(header.hash())?;
+        let epoch_first_block_hash = block_info.epoch_first_block();
+
+        if *epoch_first_block_hash == CryptoHash::default() {
+            // This is the genesis epoch. We don't have any fully finalised epoch yet.
+            return Ok(());
+        }
+
+        let epoch_first_block_info = self.epoch_manager.get_block_info(epoch_first_block_hash)?;
+        let prev_epoch_last_block_hash = epoch_first_block_info.prev_hash();
+
+        if *prev_epoch_last_block_hash == CryptoHash::default() {
+            // This is the genesis epoch. We don't have any fully finalised epoch yet.
+            return Ok(());
+        }
+        let prev_epoch_last_block_info =
+            self.epoch_manager.get_block_info(prev_epoch_last_block_hash)?;
+
+        if prev_epoch_last_block_info.epoch_id() == epoch_first_block_info.epoch_id() {
+            // Previous epoch is the genesis epoch. We don't have any fully finalised epoch yet.
```

Comment:
Is this possible? If not, log_assert() or some other way to indicate an unexpected condition, please.
The only two epochs with the same epoch_id are:
* The epoch that consists of the genesis block
* The next epoch, i.e. the first epoch at epoch_height=1

Reply:
Yes, it is possible. As you noted, it is only possible for genesis epoch, and I specifically want to avoid genesis epoch for two reasons: 
- Everyone has full genesis epoch on startup
- Next epoch has the same epoch_id and therefore it is incorrect to have that key in DB before the end of first proper epoch